### PR TITLE
Use `jax.numpy.diagonal` to evaluate `jax.numpy.trace`.

### PR DESCRIPTION
**Pull request by tillahoffmann**
The current implementation of `jax.numpy.trace` uses masking to include/exclude the elements to trace. This can be slow, especially for larger arrays. This change instead uses `jax.numpy.diagonal` to extract the elements and evaluate the trace.

The performance gain depends on the size of the input arrays but can be >100x for larger arrays.

```python
# Use lambda to treat axis and offset as static args.
jitted_old = jax.jit(lambda x: old_trace(x))
jitted_new = jax.jit(lambda x: new_trace(x))

# Run the tests for arrays of different sizes.
key = jax.random.key(18)
for shape in [(10, 10), (100, 100, 8), (1000, 1000, 17)]:
    print(shape)
    key, subkey = jax.random.split(key)
    x = jax.random.normal(subkey, shape)
    for method, f in [("old", jitted_old), ("new", jitted_new)]:
        print(method)
        # Run once to ensure compiled for this shape.
        f(x).block_until_ready()
        # Run the timings.
        %timeit f(x).block_until_ready()
```
```
(10, 10)
old
4.17 µs ± 79.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
new
3.41 µs ± 32.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

(100, 100, 8)
old
107 µs ± 906 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
new
4.36 µs ± 61.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

(1000, 1000, 17)
old
31.4 ms ± 1.21 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
new
35.4 µs ± 383 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

### CODE REVIEW

**Comment by jakevdp** [comment]
Thanks for the suggestion – it might be worth benchmarking on TPU as well. `gather` (which is the operation behind indexing) is typically very slow on TPU due to constraints of the hardware architecture, which may be why the masking implementation was chosen in the first place.

**Comment by tillahoffmann** [comment]
Interesting. Here are the results from a Colab TPU (with slightly larger trailing dimension to minimize profiling the overhead). It seems that the `diagonal`-based version still does quite well.

```
(10, 10, 500)
old
132 µs ± 2.37 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
136 µs ± 2.02 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(100, 100, 500)
old
218 µs ± 7.32 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
new
132 µs ± 7.76 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(1000, 1000, 500)
old
7.07 ms ± 25.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
new
318 µs ± 19.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Here's the full script I ran.

```python
import jax
from jax import numpy as jnp
from jax.numpy import trace as old_trace
import numpy as np


def new_trace(a, offset=0, axis1=0, axis2=1):
    return jnp.diagonal(a, offset=offset, axis1=axis1, axis2=axis2).sum(axis=-1)


# Use lambda to treat axis and offset as static args.
jitted_old = jax.jit(lambda x: old_trace(x))
jitted_new = jax.jit(lambda x: new_trace(x))

# Run the tests for arrays of different sizes.
key = jax.random.key(18)
for shape in [(10, 10, 500), (100, 100, 500), (1000, 1000, 500)]:
    print(shape)
    key, subkey = jax.random.split(key)
    x = jax.random.normal(subkey, shape)
    previous = None
    for method, f in [("old", jitted_old), ("new", jitted_new)]:
        print(method)
        # Run once to ensure compiled for this shape.
        y = f(x).block_until_ready()
        # Check the results are the same.
        if previous is not None:
          np.testing.assert_allclose(previous, y)
        # Run the timings.
        %timeit f(x).block_until_ready()
```

**Comment by hawkinsp** [comment]
Can you try dimensions 128x128 and 256x256 without a batch dimension, on colab TPU?

(I strongly suspect that on TPU within a block of size 128x128 the masking is optimal. If that were the case, then the optimal strategy is probably a hybrid of the two: gather diagonal blocks and then mask them. But it's possible XLA is smart enough to figure that out already.)

**Comment by tillahoffmann** [comment]
Looks like your hunch is right. I've included both with and without batch dimension. I suspect we may be profiling call overhead without the batch dimension because the time taken is almost constant.

**With batch dimension**

```
(32, 32, 500)
old
139 µs ± 3.18 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
123 µs ± 6.9 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(64, 64, 500)
old
161 µs ± 5.02 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
123 µs ± 5.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(128, 128, 500)
old
400 µs ± 65.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
new
940 µs ± 26.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
(256, 256, 500)
old
866 µs ± 13.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
new
1.92 ms ± 27.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
(512, 512, 500)
old
2.34 ms ± 36.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
new
7.77 ms ± 16.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
(1024, 1024, 500)
old
7.51 ms ± 26.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
new
29.1 ms ± 25 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Without batch dimension**

```
(32, 32)
old
126 µs ± 16.1 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
116 µs ± 3.6 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(64, 64)
old
126 µs ± 8.95 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
144 µs ± 16.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(128, 128)
old
112 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
136 µs ± 17.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(256, 256)
old
139 µs ± 18.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
119 µs ± 4.76 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(512, 512)
old
132 µs ± 14.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
133 µs ± 1.92 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
(1024, 1024)
old
131 µs ± 5.61 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new
165 µs ± 2.61 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
